umbraco/Umbraco-CMS
# Review: Allow Duplicate Email for Members

fixes https://github.com/umbraco/Umbraco-CMS/issues/15676

### Prerequisites



### Description

This is a redo of this PR https://github.com/umbraco/Umbraco-CMS/pull/14784. That PR went stale and the Contrib branch advanced to v14, so this is work for v13.
---------------------
Oddly have had two project in a row that the client had the business need to allow member to have duplicate email address. While there are other way to resolve this issue, want to add (re-add?) this feature to Umbraco so that it was easy to enable with out having to work out complicated alternative contact email options. So, added the option to allow member accounts to be created with duplicate emails.

The current default behavior is that members must have unique emails. While it was possible to override the ASP.NET Identity provider value for RequireUniqueEmail in startup.cs or program.cs, validators still checked if the email existed and prevented the member from being added if the email was a duplicate.

Add a new appsettings.json option to the **Umbraco:CMS:Security** section, **RequireUniqueEmailForMembers** with a default value of true. This now sets the value for **options.User.RequireUniqueEmail** in the **ConfigureMemberIdentityOptions** class.

Also added a new method to the MemberService, **GetMembersByEmail**.The **GetByEmail** method will only return FirstOrDefault. This new method allows for the retrieval of all member with the same email.

This option only allows duplicate emails for members and not users. Usernames must still be unique.

#### TEST

1. Start up a new solution.
2. Go to the Members section and create a new member.
3. Create a second new member using the same email as the one just created. Validation should prevent the creation of this member.
4. Stop the solution.
5. Add the new security option to the appsettings.json, **Umbraco:CMS:Security:RequireUniqueEmailForMembers**, set the value to **true** and save file
6. Start up a the solution, go to the Member section, and repeat step 3 - 4, Validation should still prevent the creation of the member.
7. Set **RequireUniqueEmailForMembers** to **false** and save file.
8. Start up a the solution, go to the Member section, and repeat step 3 - 4, the member should now be created.
9. Stop the solution.


---
_This item has been added to our backlog AB#47446_
